### PR TITLE
qa/tasks/cephadm: fix pulling containers from private registries

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Log in to a registry before calling cephadm pull.

Fixes: https://tracker.ceph.com/issues/65398 (maybe)

Extremely untested at this point.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests
  - [x] Tries to improve test framework

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
@dmick PTAL ^

**Comment by phlogistonjohn** [comment]
I need a private repo to test with. I don't want to mess with my account on the real quay.io and I don't have the ability to create private repos there. And I have various irrational(?) reasons I don't want to sign back up with dockerhub after recently deleting my account there.

**Comment by dmick** [comment]
See PMs, and I'll see if I can try this tonight.

**Comment by phlogistonjohn** [comment]
I was able to test this today with some additional hacks, I added patches to the PR that support these hacks but without the hacked up values used anywhere. Let me know if want them or not.

File: qa/tasks/cephadm.py [hunk 1]
```diff
@@ -2179,6 +2179,9 @@ def task(ctx, config):
     containers = config.get('containers', {})
     container_image_name = containers.get('image', container_image_name)
 
+    if 'force_image' in config:
```
Comment (dmick) [inline qa/tasks/cephadm.py:2182]: I think this does the same thing as config.image.  Here's a bit of doc I wrote that matches my experience with the existing code:

+    Set cephadm.image to supply the whole image name, including version.
+    Set cephadm.containers.image to an image tag without version, and the task
+     will use the toplevel config items sha1:, flavor:, and branch: to
+     create either :sha1, :sha1-flavor:, or :branch (default main) as a version
+

Comment (phlogistonjohn) [reply qa/tasks/cephadm.py:2182]: Interesting, thanks.  I mentioned to @adk3798 earlier today I had trouble testing the 1st patch without these hacks but I could have messed something up. If @adk3798 agrees that this is not needed I can drop this patch from the PR and then I will retest at my leisure later and if I find I still need this for some reason I'll try posting it again.

I find the code that implements all this derivation of the container name incredibly confusing to the point that I tried and failed to refactor it once before! Maybe the confusing code prevents me from understanding the intended workflow ;-)

Comment (dmick) [reply qa/tasks/cephadm.py:2182]: the code certainly could use some inline comments too as well as the above.

**Comment by phlogistonjohn** [comment]
I have dropped the patch related to the 'force_image' option as both Dan and Adam have shown that it was redundant. I think the rest of the patches accomplish something useful so I'm marking this as ready for review.

**Comment by adk3798** [comment]
jenkins test make check

**Review by adk3798 — APPROVED** [review]

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-05-09_03:09:40-orch:cephadm-wip-adk-testing-2024-05-08-1927-distro-default-smithi/

failures:
- 21 `in cluster log` type failures that are still being cleaned up, known issue
- 3 mds_upgrade_sequence failures, known issue
- 1 test failed installing `ceph-fuse` ` Status code: 503 for https://mirrors.centos.org/metalink?repo=centos-baseos-9-stream&arch=x86_64&protocol=https,http`, happens on occasion, nothing to block merging over

Overall, nothing unexpected in the run

**Comment by adk3798** [comment]
jenkins test make check

**Pull request merged**